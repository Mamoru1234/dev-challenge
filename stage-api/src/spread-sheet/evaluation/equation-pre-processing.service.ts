import { Injectable } from '@nestjs/common';
import { EquationValue, parseValue } from '../equation-value';
import { uniq } from 'lodash';
import urlRegexSafe from 'url-regex-safe';
import { LINK_VAR_PREFIX } from './evaluation.constants';

export interface PreProcessingOutput {
  value: string;
  defaultVars?: Record<string, EquationValue>;
}

function getURLs(value: string) {
  return uniq(value.match(urlRegexSafe()));
}

@Injectable()
export class EquationPreProcessingService {
  async process(value: string): Promise<PreProcessingOutput> {
    let returnValue = value.trim();
    if (!returnValue.startsWith('=')) {
      return {
        value: returnValue,
      };
    }
    const defaultVars: Record<string, EquationValue> = {};
    const urls = getURLs(value);
    for (let i = 0; i < urls.length; i++) {
      const varName = `${LINK_VAR_PREFIX}${i}`;
      const url = urls[i];
      returnValue = returnValue.replaceAll(url, varName);
      defaultVars[varName] = parseValue(url);
    }
    return {
      value: returnValue,
      defaultVars,
    };
  }
}
