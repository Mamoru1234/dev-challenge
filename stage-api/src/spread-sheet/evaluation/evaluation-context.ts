import { EquationVariablesService } from './equation-variables.service';
import { ExternalArgsFetcher } from './external-args.fetcher';

export class EvaluationContext {
  constructor(
    readonly externalArgsFetcher: ExternalArgsFetcher,
    readonly variablesService: EquationVariablesService,
  ) {}
}
