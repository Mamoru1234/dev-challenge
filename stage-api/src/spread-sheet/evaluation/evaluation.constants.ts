export const LINK_VAR_PREFIX = 'systemUrl';
