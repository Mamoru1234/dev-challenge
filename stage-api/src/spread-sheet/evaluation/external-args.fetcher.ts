import { HttpService } from '@nestjs/axios';
import { BadRequestException, Logger } from '@nestjs/common';
import { plainToInstance } from 'class-transformer';
import { firstValueFrom } from 'rxjs';
import { ExternalResourceResponse } from './dto/external-resource.response';
import { validate } from 'class-validator';
import { ExternalValueEntity } from '../../database/entity/external-value.entity';
import { Repository } from 'typeorm';

export class ExternalArgsFetcher {
  private readonly logger = new Logger(ExternalArgsFetcher.name);

  constructor(
    private readonly httpService: HttpService,
    private readonly externalValueRepository: Repository<ExternalValueEntity>,
  ) {}

  async fetch(url: string): Promise<string> {
    try {
      const existingExternal = await this.externalValueRepository.findOne({
        where: {
          url,
        },
        select: ['id', 'result'],
      });
      if (existingExternal) {
        return existingExternal.result;
      }
      const newExternal = await this.initExternalValue(url);
      return newExternal.result;
    } catch (e) {
      if (e instanceof BadRequestException) {
        throw e;
      }
      this.logger.warn('Error during fetch', e);
      throw new BadRequestException('External arg cannot be fetched');
    }
  }

  private async fetchExternalValue(url: string) {
    const { data } = await firstValueFrom(
      this.httpService.get(url, {
        maxContentLength: 2048,
      }),
    );
    const payload = plainToInstance(ExternalResourceResponse, data, {
      excludeExtraneousValues: true,
      enableImplicitConversion: true,
    });
    const errors = await validate(payload);
    if (errors.length) {
      this.logger.log('Errors validating payload', {
        payload,
        errors,
      });
      throw new BadRequestException('External resource response is not valid');
    }
    return payload;
  }

  private async subscribeOnUpdate(id: number, url: string) {
    await firstValueFrom(
      this.httpService.post(
        `${url}/subscribe`,
        {
          webhook_url: `${process.env.SERVER_BASE}/subscriptions/${id}`,
        },
        {
          maxContentLength: 1024,
        },
      ),
    );
  }

  async initExternalValue(url: string): Promise<ExternalValueEntity> {
    const { result } = await this.fetchExternalValue(url);
    const entity = await this.externalValueRepository.save({
      url,
      result,
    });
    await this.subscribeOnUpdate(entity.id, url);
    return entity;
  }
}
