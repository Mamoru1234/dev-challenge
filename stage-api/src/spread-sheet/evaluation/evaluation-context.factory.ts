import { Injectable } from '@nestjs/common';
import { EvaluationContext } from './evaluation-context';
import { ExternalArgsFetcher } from './external-args.fetcher';
import { EntityManager } from 'typeorm';
import { CellEntity } from '../../database/entity/cell.entity';
import { EquationVariablesService } from './equation-variables.service';

@Injectable()
export class EvaluationContextFactory {
  constructor(private readonly externalArgsFetcher: ExternalArgsFetcher) {}

  createContext(tx: EntityManager): EvaluationContext {
    const cellRepository = tx.getRepository(CellEntity);
    const variablesService = new EquationVariablesService(cellRepository);
    return new EvaluationContext(this.externalArgsFetcher, variablesService);
  }
}
