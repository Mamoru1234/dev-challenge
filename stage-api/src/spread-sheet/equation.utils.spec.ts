import { EquationNode, parse } from 'equation-parser';
import { evalEquation, findVariables, parseEquation } from './equation.utils';
import { BadRequestException } from '@nestjs/common';
import { EvaluationContext } from './evaluation/evaluation-context';

describe('equation utils', () => {
  describe('evaluate', () => {
    function generateCases(cases: [string, number][]) {
      for (const current of cases) {
        it(`${current[0]} = ${current[1]}`, async () => {
          expect(
            await evalEquation(
              parse(current[0]) as EquationNode,
              {},
              new EvaluationContext(null, null),
            ),
          ).toEqual(current[1]);
        });
      }
    }

    generateCases([
      ['1 + 1', 2],
      ['1 - 2 + 10', 9],
      ['2 ^ 2', 4],
      ['2 ^ 5 / 2 ^ 3', 4],
      ['min(1, 2^5, 30 * 100)', 1],
      ['MIN(1, 2^5, 30 * 100)', 1],
      ['MiN(1, 2^5, 30 * 100)', 1],
      ['AVG(2, 3, 4, 5)', 3.5],
      ['Sum(1, 5, 6, 7)', 19],
      ['Max(2, 3, sum(1, 3, 10))', 14],
    ]);
  });

  describe('findVariables', () => {
    function generateCases(cases: [string, string[]][]) {
      for (const current of cases) {
        it(`${current[0]} should find ${JSON.stringify(current[1])}`, () => {
          expect(findVariables(parse(current[0]) as EquationNode)).toEqual(
            current[1],
          );
        });
      }
    }

    generateCases([
      ['1 + 1', []],
      ['1 + var', ['var']],
      ['1 + var1 / ter + 234', ['var1', 'ter']],
      ['1 + Var1 / tEr + 234', ['var1', 'ter']],
    ]);
  });

  describe('parseEquation', () => {
    it('parse error', () => {
      expect(() => parseEquation('=a +')).toThrow(BadRequestException);
    });

    it('should return null if not equation', () => {
      const result = parseEquation('a + b');
      expect(result).toBe(null);
    });

    it('should return equation node', () => {
      const result = parseEquation('=a+b');
      expect(result).toMatchSnapshot();
    });
  });
});
