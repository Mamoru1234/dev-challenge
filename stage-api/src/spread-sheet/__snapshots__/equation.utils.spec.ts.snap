// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`equation utils parseEquation should return equation node 1`] = `
{
  "a": {
    "name": "a",
    "type": "variable",
  },
  "b": {
    "name": "b",
    "type": "variable",
  },
  "type": "plus",
}
`;
